lib/fragments.js

```diff
@@ -19,13 +19,13 @@ const buildFragments = function (document) {
   }
 
   for (const [root, fieldDefs] of Object.entries(tree)) {
-    const fields= [];
-
+    const fields = [];
     for (const { name, type } of fieldDefs) {
       let current = type;
 
-      if (type.kind === Kind.LIST_TYPE) {
-        current = type.type;
+      // traverse depth first to handle Kind.LIST_TYPE and KIND.NON_NULL_TYPE nodes
+      while (current.kind !== Kind.NAMED_TYPE) {
```

Comment:
Assuming we will always find a NAMED_TYPE?

Reply:
Great question. I believe so. Going off of the graphql-js reference implementation. `fieldDefs` are an array of FieldDefinitionNode: https://github.com/graphql/graphql-js/blob/master/src/language/ast.js#L484

The `type` field on the `FieldDefinitionNode` is a `TypeNode`: https://github.com/graphql/graphql-js/blob/master/src/language/ast.js#L414 which can be one of three things: `NamedTypeNode`, `ListTypeNode`, `NonNullTypeNode`.

From there it appears as though `ListTypeNode` and `NonNullTypeNode` wrap a `NamedTypeNode`. That is:

* `ListTypeNode` has a type field which is a `NamedTypeNode`
* `NonNullTypeNode` has a type field which can be either a `ListTypeNode` or a `NamedTypeNode`

Do you think it'd be a good idea to break out of that loop if `current` is set to undefined? From what I can tell you will have to have a `NAMED_TYPE`